pkg/gadgets/trace/tcpconnlat/tracer/tracer.go

```diff
@@ -0,0 +1,204 @@
+	MountnsMap *ebpf.Map
+	minLatency uint64
+}
+
+type Tracer struct {
+	config        *Config
+	eventCallback func(*types.Event)
+
+	objs   tcpconnlatObjects
+	links  []link.Link
+	reader *perf.Reader
+}
+
+func (t *Tracer) close() {
+	for _, link := range t.links {
+		gadgets.CloseLink(link)
+	}
+	t.links = nil
+	t.objs.Close()
+}
+
+func (t *Tracer) install() error {
+	var err error
+	var l link.Link
+
+	spec, err := loadTcpconnlat()
+	if err != nil {
+		return fmt.Errorf("loading ebpf spec: %w", err)
+	}
+
+	gadgets.FixBpfKtimeGetBootNs(spec.Programs)
+
+	mapReplacements := map[string]*ebpf.Map{}
+	filterByMntNs := false
+
+	if t.config.MountnsMap != nil {
+		filterByMntNs = true
+		mapReplacements["mount_ns_filter"] = t.config.MountnsMap
+	}
+
+	consts := map[string]interface{}{
+		"filter_by_mnt_ns": filterByMntNs,
+		"targ_min_us":      t.config.minLatency,
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil {
+		return fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	opts := ebpf.CollectionOptions{
+		MapReplacements: mapReplacements,
+	}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		return fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	l, err = link.Kprobe("tcp_v4_connect", t.objs.IgTcpcV4CoE, nil)
+	if err != nil {
+		return fmt.Errorf("attaching program: %w", err)
+	}
+	t.links = append(t.links, l)
+
+	l, err = link.Kprobe("tcp_v6_connect", t.objs.IgTcpcV6CoE, nil)
+	if err != nil {
+		return fmt.Errorf("attaching program: %w", err)
+	}
+	t.links = append(t.links, l)
+
+	l, err = link.Kprobe("tcp_rcv_state_process", t.objs.IgTcpRsp, nil)
+	if err != nil {
+		return fmt.Errorf("attaching program: %w", err)
+	}
+	t.links = append(t.links, l)
+
+	reader, err := perf.NewReader(t.objs.tcpconnlatMaps.Events, gadgets.PerfBufferPages*os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("creating perf ring buffer: %w", err)
+	}
+	t.reader = reader
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	for {
+		record, err := t.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				// nothing to do, we're done
+				return
+			}
+
+			msg := fmt.Sprintf("Error reading perf ring buffer: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if record.LostSamples > 0 {
+			msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+			t.eventCallback(types.Base(eventtypes.Warn(msg)))
+			continue
+		}
+
+		bpfEvent := (*tcpconnlatEvent)(unsafe.Pointer(&record.RawSample[0]))
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type:      eventtypes.NORMAL,
+				Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
+			},
+			WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MntnsId},
+			Pid:           bpfEvent.Tgid,
+			Comm:          gadgets.FromCString(bpfEvent.Comm[:]),
+			Dport:         gadgets.Htons(bpfEvent.Dport),
+			Sport:         bpfEvent.Lport,
+			Latency:       time.Duration(int64(bpfEvent.DeltaUs * 1000)),
```

Comment:
The latency is initially calculated in nanoseconds, then converted to microseconds, then back to nanoseconds here. Could we keep nanoseconds everywhere to avoid both conversions?